### PR TITLE
PSQLODBC Datatype Test for sys.bigint

### PR DESCRIPTION
**Pull request by ColinKYuen**
### Description

Supersedes #3 

Adds ODBC Datatype Test for `sys.bigint`
- [x] Create Table
- [x] Insert valid values
- [x] Insert invalid values
- [x] Update table with valid values
- [x] Update table with invalid values
- [x] Arithmetic Operators
- [x] Comparison Operators
- [x] Comparison Functions (Min/Max)
- [x] Table View
- [x] Unique Constraint
- [x] Composite Key

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by ColinKYuen** [comment]
Supersedes #3

File: test/odbc/psqlodbc/test_pg_datatype_bigint.cpp [hunk 1]
```diff
@@ -0,0 +1,1016 @@
...
+#include <time.h>
+using std::pair;
+
+const string TABLE_NAME = "master_dbo.bigint_table_odbc_test";
+const string COL1_NAME = "pk";
+const string COL2_NAME = "data";
+const string DATATYPE_NAME = "sys.bigint";
+const string VIEW_NAME = "master_dbo.bigint_view_odbc_test";
+vector<pair<string, string>> TABLE_COLUMNS = {
+  {COL1_NAME, "INT PRIMARY KEY"},
```
Comment (faviansamatha) [inline test/odbc/psqlodbc/test_pg_datatype_bigint.cpp:17]: Why was this changed to "INT"? The original tests used Big int datatype for primary key?

File: test/odbc/psqlodbc/test_pg_datatype_bigint.cpp [hunk 2]
```diff
@@ -0,0 +1,1016 @@
...
+
+  odbcHandler.ExecQuery(DropObjectStatement("TABLE", TABLE_NAME));
+}
+
+TEST_F(PSQL_DataTypes_BigInt, Update_Fail) {
+  const int PK_INSERTED = 1;
+  const string DATA_INSERTED = "12345";
+  const string DATA_UPDATED_VALUE = "9223372036854775808"; // Over max
+  
+
```
Comment (faviansamatha) [inline test/odbc/psqlodbc/test_pg_datatype_bigint.cpp]: nit: extra line

File: test/odbc/psqlodbc/test_pg_datatype_bigint.cpp [hunk 3]
```diff
@@ -0,0 +1,1016 @@
...
+    "9223372036854775807",
+    "123456789"
+  };
+  const int NUM_OF_DATA = DATA_UPDATED_VALUES.size();
+
+  const string INSERT_STRING = "(" + std::to_string(PK_INSERTED) + "," + DATA_INSERTED + ")";
+  const string UPDATE_WHERE_CLAUSE = COL1_NAME + " = " + std::to_string(PK_INSERTED);
+
+  const int PK_BYTES_EXPECTED = 4;
+  const int DATA_BYTES_EXPECTED = 8;
```
Comment (faviansamatha) [inline test/odbc/psqlodbc/test_pg_datatype_bigint.cpp]: `PK_BYTES_EXPECTED` and `DATA_BYTES_EXPECTED` are used in several locations. Should we consider to make this global? Or maybe a more general case and name it `BIGINT_EXPECTED_BYTES` and `INT_EXPECTED_BYTES`?

File: test/odbc/psqlodbc/test_pg_datatype_bigint.cpp [hunk 4]
```diff
@@ -0,0 +1,1016 @@
...
+  RETCODE rcode;
+  OdbcHandler odbcHandler(Drivers::GetDriver(ServerType::PSQL));
+
+  const vector<string> INSERTED_PK = {
+    "-73708",
+    "123",
+    "233"
+  };
+
+  const vector<string> INSERT_DATA = {
```
Comment (faviansamatha) [inline test/odbc/psqlodbc/test_pg_datatype_bigint.cpp]: nit: Let's be a bit more consistent with our naming and call this `INSERTED_DATA` since we have a variable named `INSERTED_PK`.

File: test/odbc/psqlodbc/test_pg_datatype_bigint.cpp [hunk 5]
```diff
@@ -0,0 +1,1016 @@
...
+    expected_results[i].push_back(abs(data_1));
+  }
+
+  long long int col_results[NUM_OF_OPERATIONS];
+  SQLLEN col_len[NUM_OF_OPERATIONS];
+  vector<tuple<int, int, SQLPOINTER, int, SQLLEN *>> BIND_COLUMNS = {};
+
+  // initialization for BIND_COLUMNS
+  for (int i = 0; i < NUM_OF_OPERATIONS; i++) {
+    tuple<int, int, SQLPOINTER, int, SQLLEN *> tuple_to_insert(i + 1, SQL_C_SBIGINT, (SQLPOINTER)&col_results[i], 0, &col_len[i]);
```
Comment (faviansamatha) [inline test/odbc/psqlodbc/test_pg_datatype_bigint.cpp:438]: Some of these operators in Postgres don't return type bigint (e.g., square root or cube root). Do we still want to bind it to SQL_C_SBIGINT? I assume that it would round it down to the nearest int, but we're also losing decimal places from these functions. 

if I were a user who wanted to know the square root of 2 and 2 was of type big int, I wouldn't be satisfied with the number 1.

File: test/odbc/psqlodbc/test_pg_datatype_bigint.cpp [hunk 6]
```diff
@@ -0,0 +1,1016 @@
...
+
+    min_expected = std::min(curr, min_expected);
+    max_expected = std::max(curr, max_expected);
+  }
+  expected_results.push_back(min_expected);
+  expected_results.push_back(max_expected);
+  expected_results.push_back(sum);
+  expected_results.push_back(sum / NUM_OF_DATA);
+  
+
```
Comment (faviansamatha) [inline test/odbc/psqlodbc/test_pg_datatype_bigint.cpp:622]: nit: extra line

File: test/odbc/psqlodbc/test_pg_datatype_bigint.cpp [hunk 7]
```diff
@@ -0,0 +1,1016 @@
...
+  odbcHandler.ExecQuery(DropObjectStatement("VIEW", VIEW_NAME));
+
+  odbcHandler.CloseStmt();
+  odbcHandler.ExecQuery(DropObjectStatement("TABLE", TABLE_NAME));
+}
+
+TEST_F(PSQL_DataTypes_BigInt, Table_Unique_Constraints) {
+  const vector<pair<string, string>> TABLE_COLUMNS = {
+    {COL1_NAME, "INT PRIMARY KEY"},
+    {COL2_NAME, DATATYPE_NAME + " UNIQUE"}
```
Comment (faviansamatha) [inline test/odbc/psqlodbc/test_pg_datatype_bigint.cpp:759]: Would naming the unique constraint be helpful?

File: test/odbc/psqlodbc/test_pg_datatype_bigint.cpp [hunk 8]
```diff
@@ -0,0 +1,1016 @@
...
+
+  const string PK_QUERY =
+    "SELECT C.COLUMN_NAME FROM "
+    "INFORMATION_SCHEMA.TABLE_CONSTRAINTS T "
+    "JOIN INFORMATION_SCHEMA.CONSTRAINT_COLUMN_USAGE C "
+    "ON C.CONSTRAINT_NAME=T.CONSTRAINT_NAME "
+    "WHERE "
+    "C.TABLE_NAME='" +
+    TABLE_NAME.substr(TABLE_NAME.find('.') + 1, TABLE_NAME.length()) + "' "
+    "AND T.CONSTRAINT_TYPE='UNIQUE'";
```
Comment (faviansamatha) [inline test/odbc/psqlodbc/test_pg_datatype_bigint.cpp:811]: What is this trying to do? It's called PK_QUERY, but is it trying to get the unique key? Also, can't we use pg_constraint for this instead? Would be better than going through information_schema since it could be OID-based.

File: test/odbc/psqlodbc/test_pg_datatype_bigint.cpp (hunk 1 above)
Comment (ColinKYuen) [reply test/odbc/psqlodbc/test_pg_datatype_bigint.cpp:17]: Consistency with other tests.
The Primary key in most tests are inserted using a for loop's index and with some tests, using the test type wouldn't work eg DateTime

Comment (faviansamatha) [reply test/odbc/psqlodbc/test_pg_datatype_bigint.cpp:17]: For some of those types yes, but would be nice to test the primary key constraint of some of these types as well, but not a hard requirement.

**Comment by ColinKYuen** [comment]
Superseded by #28
